### DISCUSSION TITLE
Best practices with displaying content from a changing data source

### DISCUSSION THREAD

**Discussion by papr** [Textualize/textual#509] [Q&A]
Dear Textualize team,

Let me start by thanking you for this great project! I am already a big Rich fan and have just started picking up Textual.

While prototyping a small project, I encountered a few use cases where I wasn't sure what the best practices would look like regarding their implementation.

Thank you very much in advance for considering my questions 🙏🏻 

## Context

The project is a small command line utility that interacts with another device in the local network. The CLI allows the user to specify the address of the device manually. Alternatively, the device announces its service via Bonjour/mDNS. So if the user does not provide an address, I want to show a small textual app that lists all available devices and lets the user select one. This is what my current prototype ([source code](https://gist.github.com/papr/8c33064306e026545ad6ee2cec902879)) looks like:
![Screenshot 2022-05-12 at 15 51 30](https://user-images.githubusercontent.com/168390/168307755-8357b226-cd6d-44c4-bbfa-723c3084a736.png)

These are the situations where I would appreciate your input regarding best practices:

### Listing available devices

I want to list all devices with a fixed height, in a table-like manner. The current implementation uses a table that redraws from scratch on every key input, keeping track of the selected index. Looking at DockView and the focus/hover functionality of widgets, this feels like reinventing the wheel.

#### Scrollable DockView

Ideally, the view would scroll in case of overflow. My first approach was to wrap a DockView in a ScrollView. But in that case, nothing gets rendered.

The example in [this tweet](https://twitter.com/willmcgugan/status/1522591188372627456) looks like exactly what I need but I did not find the source code yet.

#### Reloading DockView content 

What is the best practice to reload the content of a dock view? Should I simply regenerate the whole DockView?

### Shutdown

I want to clean up the network resources on shutdown. I saw the `event.Shutdown` class but `MyApp.on_shutdown` does not seem to be called on shutdown. My current approach is to overwrite `App.shutdown()` which does not feel clean.

### Returning information on shutdown

Once the user selected a device, I would like to stop the textual app and use the selection for the actual command line utity. What would be the best practice to return a status code/information from the application? Something as easy as a custom attribute containing the result that can be accessed after shutdown?

**Comment by HiPeople21** [comment]
> The example in [this tweet](https://twitter.com/willmcgugan/status/1522591188372627456) looks like exactly what I need but I did not find the source code yet.

I'm pretty sure the source code is in the `css` branch, in the `uber.py` file in the `sandbox` folder.

**Reply by papr** [reply]
@HiPeople21 Thank you for the response! I have yet to explore the CSS branch/functionality.
